pkg/ruler/api.go

```diff
@@ -268,3 +284,241 @@ func (r *Ruler) alerts(w http.ResponseWriter, req *http.Request) {
+	}
+
+	namespace, err := url.PathUnescape(namespace)
+	if err != nil {
+		return "", err
+	}
+
+	return namespace, nil
+}
+
+// parseGroupName parses the group name from the provided set of params, in this
+// api these params are derived from the url path
+func parseGroupName(params map[string]string) (string, error) {
+	groupName, exists := params["groupName"]
+	if !exists {
+		return "", ErrNoGroupName
+	}
+
+	groupName, err := url.PathUnescape(groupName)
+	if err != nil {
+		return "", err
+	}
+
+	return groupName, nil
+}
+
+func parseRequest(req *http.Request, requireNamespace, requireGroup bool) (string, string, string, error) {
+	id, err := user.ExtractOrgID(req.Context())
+	if err != nil {
+		return "", "", "", user.ErrNoOrgID
+	}
+
+	vars := mux.Vars(req)
+
+	namespace, err := parseNamespace(vars)
+	if err != nil {
+		if err != ErrNoNamespace || requireNamespace {
+			return "", "", "", err
+		}
+	}
+
+	group, err := parseGroupName(vars)
+	if err != nil {
+		if err != ErrNoGroupName || requireGroup {
+			return "", "", "", err
+		}
+	}
+
+	return id, namespace, group, nil
+}
+
+func (r *Ruler) listRules(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+
+	userID, namespace, _, err := parseRequest(req, false, false)
+	if err != nil {
+		respondError(logger, w, err.Error())
+	}
+
+	level.Debug(logger).Log("msg", "retrieving rule groups with namespace", "userID", userID, "namespace", namespace)
+	rgs, err := r.store.ListRuleGroups(req.Context(), userID, namespace)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(logger).Log("msg", "retrieved rule groups from rule store", "userID", userID, "num_namespaces", len(rgs))
+
+	if len(rgs) == 0 {
+		level.Info(logger).Log("msg", "no rule groups found", "userID", userID)
+		http.Error(w, ErrNoRuleGroups.Error(), http.StatusNotFound)
+		return
+	}
+
+	formatted := rgs.Formatted()
+	marshalAndSend(formatted, w, logger)
+}
+
+func (r *Ruler) getRuleGroup(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, namespace, groupName, err := parseRequest(req, true, true)
+	if err != nil {
+		respondError(logger, w, err.Error())
+	}
+
+	rg, err := r.store.GetRuleGroup(req.Context(), userID, namespace, groupName)
+	if err != nil {
+		if err == store.ErrGroupNotFound {
+			http.Error(w, err.Error(), http.StatusNotFound)
+			return
+		}
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	formatted := store.FromProto(rg)
+	marshalAndSend(formatted, w, logger)
+}
+
+func (r *Ruler) createRuleGroup(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, namespace, _, err := parseRequest(req, true, false)
+	if err != nil {
+		respondError(logger, w, err.Error())
+	}
+
+	payload, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		level.Error(logger).Log("err", err.Error())
```

Comment:
There are multiple instances of log message like this, without providing any other message or context. I think some extra details could be useful for debugging.

Also, why call `err.Error()`, instead of passing `err` directly?

Reply:
I'll update these log lines to be more descriptive